File: amazon/array/0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int water=0;
        int leftMax=0;
        int rightMax=0;

        int l=0; int r= n-1;

        while(l<r){
            if(height[l]<height[r]){
                if(leftMax<=height[l]){
                    leftMax= height[l];
                }else{
                    water+= leftMax-height[l];
                }
                l++;
            }else{
                if(rightMax<=height[r]){
                    rightMax= height[r];
                }else{
                    water+= rightMax-height[r];
                }
                r--;
            }
        }

        return water;

    }
}

File: tesla/array/0042-trapping-rain-water/0042-trapping-rain-water.ts
function trap(height: number[]): number {
    /***
        first height or water contained at i 
        waterHeight[i] = Math.min(leftmaxheight, rightmaxheohgt)- height[i]

     */
    let total : number = 0;
    let waterheight_i : number = 0;
    let leftmax : number[] = [];
    let rightmax : number[] = [];
    leftmaxheight(height,leftmax);
    rightmaxheight(height,rightmax);

    for(let i=0; i<height.length; i++){
        if(i==0 || i==height.length-1){
            continue;
        }
        waterheight_i= Math.min(leftmax[i], rightmax[i])- height[i];
        total+=waterheight_i;
    }

    return total;
};


function leftmaxheight(height: number[],leftmax: number[]): number[]{
    let maxH: number =0;
    for(let i=0; i<height.length ; i++){
        maxH= Math.max(height[i],maxH);
        leftmax[i]=maxH;
    }

    return leftmax;
}

function rightmaxheight(height: number[],rightmax: number[]): number[]{
    let maxH: number =0;
   
    for(let i=height.length-1; i>=0 ; i--){
        maxH= Math.max(height[i],maxH)
        rightmax[i]=maxH;
    }

    return rightmax;
}